### PR TITLE
Fixes #29257 - Remove x-editable-rails

### PR DESCRIPTION
**Pull request by xprazak2**
Moves the setting inline edit to React using Foreman Modal and Formik form.
This allows us to drop x-editable-rails gem and get rid of one more jquery dependency.
Settings loading is done by helper component, that should be temporary
Loading should be done through API once we move the whole Setting page to React.

### CODE REVIEW

File: app/views/settings/index.html.erb
Comment (ezr-ondrej) [inline app/views/settings/index.html.erb:8]: Hmm this component has no UI at all right? It's here just to load the settings.
Would it make sense for it to be part of the setting table or if not (I can see why it's easier to do it for all settings at once) should we introduce a general component `SaveMyJsonBlobIntoRedux`?

File: app/views/settings/index.html.erb
Comment (ezr-ondrej) [inline app/views/settings/index.html.erb:28]: I :heart: this approach compared to having form component for each setting :+1:

File: webpack/assets/javascripts/react_app/components/SettingUpdateModal/components/SettingValueField.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/SettingUpdateModal/components/SettingValueField.js:10]: Still not ideal, but I believe good enough (all the convoluted statements are hidden in this component, so would be easier to refactor in the future)

File: webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableReducer.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableReducer.js]: weird name and weird placement. this is in `SettingTableReducers` right? but it's shared with the update form.
Ideal state would be, in the shared data of all settings table just noting `'editing': <id>` of the setting that should get edited. The Wrapped form would check on that state attribute, and if the id is set, it would get the proper setting from the store.

File: webpack/assets/javascripts/react_app/components/SettingsTable/index.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/SettingsTable/index.js]: we could just set the `editing: <id>` by some action, and the modal would pick up on that, so we won't need to call `setModalOpen()` explicitly.

File: webpack/assets/javascripts/react_app/components/common/forms/InputField.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/common/forms/InputField.js]: Do you use this component?

File: webpack/assets/javascripts/react_app/components/common/forms/Select.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/common/forms/Select.js]: Nice refactor :+1:
Should we maybe extract it to a separate PR?
You could even use the same ticket as we've used for #7700 that introduced those methods.

File: app/views/settings/index.html.erb
Comment (ezr-ondrej) [reply app/views/settings/index.html.erb:8]: Oh, this would probably get droped after we will manage to get this to a React Page, right?

**Review by ezr-ondrej — COMMENTED** [review]
In general, I love this, only the modal opening is bit convoluted, but I belive that can be fixed easily.

**Comment by ezr-ondrej** [comment]
Also, this https://github.com/theforeman/foreman/blob/bed53cf2cb7f0e10927ab913774457f577ee28c1/app/assets/config/manifest.js#L7 needs to get erased.

File: app/views/settings/index.html.erb
Comment (xprazak2) [reply app/views/settings/index.html.erb:8]: Exactly, I do not think we want to encourage this as a pattern.

File: webpack/assets/javascripts/react_app/components/SettingsTable/index.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/SettingsTable/index.js]: We use hooks to open or close modals, they have their own slice of redux store:

redux-modal

ForemanModal does not currently support the behaviour that you describe.

File: webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableReducer.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableReducer.js]: I moved the state, but I kept storing the whole setting - I already have access to the setting object and I know I will need it later, that will save me an additional filtering of settings from store.

**Comment by xprazak2** [comment]
I extracted https://github.com/theforeman/foreman/pull/7851, I'll rebase on it once it is merged.

**Comment by ezr-ondrej** [comment]
I think the `SettingValueSelection` test failure is related, apart of it I think it's good step forward and I would like to get it in `2.2`.
Do you think we will be able to finish this and switch to the Setting react page next week?

**Comment by xprazak2** [comment]
Updated with a test fix.

**Comment by ezr-ondrej** [comment]
Should we send some notice to katello, that their test is going to fail after merge?

**Comment by xprazak2** [comment]
Done: https://github.com/Katello/katello/pull/8857

File: app/helpers/settings_helper.rb
Comment (ezr-ondrej) [inline app/helpers/settings_helper.rb]: This is using the unencrypted value, so even the encrypted settings are readable now.

**Review by ezr-ondrej — CHANGES_REQUESTED** [review]
Overall looks great :+1:
I've found one issue though. We are exposing the encrypted settings now :(

File: webpack/assets/javascripts/react_app/components/SettingRecords/SettingRecordsSelectors.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/SettingRecords/SettingRecordsSelectors.js]: Where using `map/find` in a selector it usually makes sense to memorize them to avoid unnecessary calculations.

**Comment by xprazak2** [comment]
I added memoized selectors and encrypted settings are now hidden.

**Review by ezr-ondrej — COMMENTED** [review]
Wouldn't it be easier just to return encrypted value from the server?
The values are still accessible for (highly) experienced users through the javascript console, right?

**Comment by xprazak2** [comment]
You are right, there would still be ways to get to the values. API just returns a [placeholder](https://github.com/theforeman/foreman/blob/develop/app/models/concerns/hidden_value.rb#L4) for encrypted settings, which complicates updating an existing value. Let me see if I can find a way to deal with this.

**Comment by ezr-ondrej** [comment]
> Let me see if I can find a way to deal with this.

Currently we can't edit the value, so I think if we would manage opening the form with empty value on every edit, I think it would be sufficient.
Lets discuss this on the next maintainers meeting?

**Comment by xprazak2** [comment]
I agree showing an empty field is probably the best solution here:


encrypted-field

**Review by ezr-ondrej — COMMENTED** [review]
We need a rebase (I can still see the helper changes).
And we now show the value in the table for collection setting (instead of the label).
But I can see it's not easy to change, so I don't insist, just if you'd have an idea how to do it easily.

**Comment by ezr-ondrej** [comment]
The failures are related, but I guess easy to fix (remove the test).

File: webpack/assets/javascripts/react_app/components/SettingRecords/__tests__/SettingRecords.fixtures.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/SettingRecords/__tests__/SettingRecords.fixtures.js]: nit
```suggestion
    value: '****',
```

**Review by ezr-ondrej — CHANGES_REQUESTED** [review]
~~For some reason it doesn't preselect the setting value in select form~~ :thinking: 
EDIT: ignore me, that was my setup.

File: webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableHelpers.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableHelpers.js:107]: I'm missing a simmilar formatter for non-grouped options :thinking:
(eg `default_locale`, `default_timezone`)

File: webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableHelpers.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableHelpers.js:32]: I don't really get this :thinking:

File: app/helpers/settings_helper.rb
Comment (ezr-ondrej) [inline app/helpers/settings_helper.rb:65]: The `full_name` is not being translated anywhere else IIUC.
```suggestion
      :full_name => setting.full_name.nil? ? setting.name : _(setting.full_name),
```

File: webpack/assets/javascripts/react_app/components/SettingsTable/components/SettingName.js
Comment (ezr-ondrej) [inline webpack/assets/javascripts/react_app/components/SettingsTable/components/SettingName.js]: Alternatively to ^^ we could
```suggestion
  <React.Fragment>{setting.fullName ? __(setting.fullName) : setting.name}</React.Fragment>
```

**Comment by ezr-ondrej** [comment]
While testing more thoroughly, it seems the `SettingsController#update` differs from `API::SettingsController#update`.
I've noticed it's not possible to nilify setting through the api. As well I'd guess it'll be impossible to set numeric values to string settings (should be just edge cases, but it can be needed).

File: webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableHelpers.js
Comment (xprazak2) [reply webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableHelpers.js:32]: It transforms keys of the select options to proper case:

```js
console.log(selectValues)

[
  { value: null, label: 'Select an owner' },
  {
    group_label: 'Users',
    children: [
      { label: 'Admin', value: '4-Users' }
    ],
  },
 {
    group_label: 'Usergroups',
    children: [
      { label: 'Default Usergroup', value: '5-Usergroups' }
    ],
  }
]
```

**Comment by xprazak2** [comment]
I fixed formatting for hash selection and added translation for full name.

I am not sure which settings you tried to set empty values for, I was able to set and remove text for the `login_text` setting.

Numeric values for string settings are out of scope for these changes.

File: webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableHelpers.js
Comment (ezr-ondrej) [reply webpack/assets/javascripts/react_app/components/SettingsTable/SettingsTableHelpers.js:32]: Ah! I see `group_label => groupLabel` ^_^

**Comment by ezr-ondrej** [comment]
Please just drop the tests for the `SettingsController#update` and I think we are good to go :+1:

**Comment by xprazak2** [comment]
Updated.

**Review by ezr-ondrej — APPROVED** [review]
Thanks @xprazak2 awesome improvement! :+1: 

Thanks @sharvit for reviewing and @ares for noticing it's definitely demo worthy :)

**Comment by ezr-ondrej** [comment]
@tbrisker I've created a packaging PR, just as a reminder https://github.com/theforeman/foreman-packaging/pull/5842
But I believe that droping is automated on release? :thinking:

**Comment by tbrisker** [comment]
Since this is touching the UI I'd be happy to wait for integration tests to work again before merging, just to make sure that nothing broke. Also i think the katello test failure might be related?

**Comment by ezr-ondrej** [comment]
> Since this is touching the UI I'd be happy to wait for integration tests to work again before merging, just to make sure that nothing broke. Also i think the katello test failure might be related?

@tbrisker well depends on how fast we can fix it and currently it doesn't depend on us, the tests were passing before the last commit, so I don't think we need to block it.
There is a PR to Katello for fixing the test as commented in https://github.com/theforeman/foreman/pull/7775#issuecomment-665720181

**Review by tbrisker — APPROVED** [review]
Packaging-wise this is good to go, it's just dropping a dependency.

**Pull request merged**
